web-common/src/features/dashboards/time-dimension-details/charts/utils.ts

```diff
@@ -22,37 +29,57 @@ export function getVegaSpecForTDD(
   chartType: TDDAlternateCharts,
   expandedMeasureName: string,
   measureLabel: string,
+  isTimeComparison: boolean,
   isDimensional: boolean,
   dimensionName: string | undefined,
   comparedValues: (string | null)[] | undefined,
 ): VisualizationSpec {
-  const temporalFields = [{ name: "ts", label: "Time" }];
-  const measureFields = [{ name: expandedMeasureName, label: measureLabel }];
-  const nominalFields = [
-    {
-      name: "dimension",
-      label: dimensionName || "dimension",
-      values: comparedValues,
-    },
+  const temporalFields: ChartField[] = [{ name: "ts", label: "Time" }];
+  const measureFields: ChartField[] = [
+    { name: expandedMeasureName, label: measureLabel },
   ];
 
+  let nominalFields: ChartField[] = [];
+  if (isDimensional) {
+    nominalFields = [
+      {
+        name: "dimension",
+        label: dimensionName || "dimension",
+        values: comparedValues,
+      },
+    ];
+  } else if (isTimeComparison) {
+    nominalFields = [
+      {
+        name: "key",
+        label: "Comparing",
+        values: [expandedMeasureName, `comparison\\.${expandedMeasureName}`],
+      },
+    ];
+    temporalFields[0].tooltipName = "time";
+    measureFields[0].name = "measure";
+  }
+
   const builderChartType = TDDChartMap[chartType];
 
   const spec = buildVegaLiteSpec(
     builderChartType,
     temporalFields,
     measureFields,
-    isDimensional ? nominalFields : [],
+    nominalFields,
```

Comment:
**Optimization Issue**: The refactoring of 'getVegaSpecForTDD' introduces multiple conditional checks and array manipulations within the function, which can lead to inefficiencies, especially with larger datasets or frequent calls. The use of conditional logic to define 'nominalFields' and modifications to 'temporalFields' and 'measureFields' based on 'isTimeComparison' can be optimized. <br> **Fix**: Consider simplifying the logic by separating concerns and possibly splitting the function into smaller, more focused functions. This can improve maintainability and potentially enhance performance by avoiding unnecessary computations and memory usage. <br> **Code Suggestion**: 
 ```
 -  let nominalFields: ChartField[] = [];
 -  if (isDimensional) {
 -    nominalFields = [
 -      {
 -        name: "dimension",
 -        label: dimensionName || "dimension",
 -        values: comparedValues,
 -      },
 -    ];
 -  } else if (isTimeComparison) {
 -    nominalFields = [
 -      {
 -        name: "key",
 -        label: "Comparing",
 -        values: [expandedMeasureName, 'comparison\\\\.${expandedMeasureName}'],
 -      },
 -    ];
 -    temporalFields[0].tooltipName = "time";
 -    measureFields[0].name = "measure";
 -  }
 +  nominalFields = isDimensional ? [{ name: "dimension", label: dimensionName || "dimension", values: comparedValues }] : [];
 +  if (isTimeComparison) {
 +    nominalFields.push({ name: "key", label: "Comparing", values: [expandedMeasureName, 'comparison\\\\.${expandedMeasureName}'] });
 +    temporalFields[0].tooltipName = "time";
 +    measureFields[0].name = "measure";
 +  }
 ```